### PR TITLE
Update variable / class naming conventions

### LINKED ISSUE #572: Update variable / class naming conventions
Proposal is to normalise several class names and property names.

Related to #317

- [ ]  `RelationshipAttribute.DependentType` and `.PrincipicalType` is not adequate. Replace with something like `LeftType` vs `RightType` or  `HostType` vs `RelatedType`.
- [ ] rename `ContextEntity` to `ResourceContext`, so that it is consistent with `ResourceDefinition` and `ResourceHooks` 
- [ ] `ContextEntity.EntityType` --> `.ResourceType`  and current `.ResourceType` --> `.ResourceDefinitionType`. 
- [ ] `EntityResourceService` --> `DefaultResourceService`

### PR DESCRIPTION
**Pull request by maurei**
Closes #572 and #317

- [x]  style: rename repository and service layer: DefaultResourceRepository and DefaultResourceService and all related interfaces (see #317)
- [x]  style: rename ContextEntity --> ResourceContext (also see #317), and rename every reference to a resource to Resource (EntityType --> ResourceType, TEntity --> TResource)
- [x]  style: rename: DependentType, PrincipalType --> RightType, LeftType. Usage of DependentType and Principal type as inspired by EF core was not technically adequate.
- [x]  chore: remove redundant bindings throughout entire codebase
- [x]  refactor: regroup hook container and executors in to pipeline-specific interfaces
- [x]  style: rename IGenericProcessor to IHasManyThroughUpdateHelper (see #317)
- [x]  style: rename IGenericProcessorFactory to IGenericServiceFactory. This is more up to date because it is not only used to instantiate operation processors, but any generic service can be instantiated with it.

### CODE REVIEW

File: src/JsonApiDotNetCore/Data/IDbContextResolver.cs [hunk 1]
```diff
@@ -7,8 +7,8 @@ public interface IDbContextResolver
     {
         DbContext GetContext();
 
-        [Obsolete("Use DbContext.Set<TEntity>() instead", error: true)]
-        DbSet<TEntity> GetDbSet<TEntity>() 
-            where TEntity : class;
+        [Obsolete("Use DbContext.Set<TResource>() instead", error: true)]
```
Comment (wisepotato) [inline src/JsonApiDotNetCore/Data/IDbContextResolver.cs]: reason for still having this?

File: src/JsonApiDotNetCore/QueryParameterServices/SparseFieldsService.cs [hunk 2]
```diff
@@ -71,7 +71,7 @@ public virtual void Parse(KeyValuePair<string, StringValues> queryParameter)
 
                 var relationship = _requestResource.Relationships.SingleOrDefault(a => a.Is(navigation));
                 if (relationship == null)
-                    throw new JsonApiException(400, $"\"{navigation}\" in \"fields[{navigation}]\" is not a valid relationship of {_requestResource.EntityName}");
+                    throw new JsonApiException(400, $"\"{navigation}\" in \"fields[{navigation}]\" is not a valid relationship of {_requestResource.ResourceName}");
```
Comment (wisepotato) [inline src/JsonApiDotNetCore/QueryParameterServices/SparseFieldsService.cs]: you're using double quotes inside the message, i would suggest using single quotes, as there's really no need to have double quotes, the messages will not have quoted double quotes in that way.

File: src/JsonApiDotNetCore/QueryParameterServices/SparseFieldsService.cs [hunk 3]
```diff
@@ -83,10 +83,10 @@ public virtual void Parse(KeyValuePair<string, StringValues> queryParameter)
         /// </summary>
         private void RegisterRelatedResourceField(string field, RelationshipAttribute relationship)
         {
-            var relationProperty = _resourceGraph.GetContextEntity(relationship.DependentType);
+            var relationProperty = _resourceGraph.GetResourceContext(relationship.RightType);
             var attr = relationProperty.Attributes.SingleOrDefault(a => a.Is(field));
             if (attr == null)
-                throw new JsonApiException(400, $"'{relationship.DependentType.Name}' does not contain '{field}'.");
+                throw new JsonApiException(400, $"'{relationship.RightType.Name}' does not contain '{field}'.");
```
Comment (wisepotato) [inline src/JsonApiDotNetCore/QueryParameterServices/SparseFieldsService.cs]: seeing as you use single quotes here as well...

File: test/JsonApiDotNetCoreExampleTests/Acceptance/Spec/SparseFieldSetTests.cs [hunk 4]
```diff
@@ -73,7 +73,7 @@ public async Task Can_Select_Sparse_Fieldsets()
             var query = _dbContext
                 .TodoItems
                 .Where(t => t.Id == todoItem.Id)
-                .Select(_resourceGraph.GetAttributes<TodoItem>(e => new { e.Id, e.Description, e.CreatedDate, e.AchievedDate } ).ToList());
+                .Select(_resourceGraph.GetAttributes<TodoItem>(e => new { e.Id, e.Description, e.CreatedDate, e.AchievedDate } ).ToArray());
```
Comment (wisepotato) [inline test/JsonApiDotNetCoreExampleTests/Acceptance/Spec/SparseFieldSetTests.cs]: so to not have to do `ToArray all the time, why not expect an `IEnumerable`?

File: src/JsonApiDotNetCore/Data/IDbContextResolver.cs (hunk 1 above)
Comment (maurei) [reply src/JsonApiDotNetCore/Data/IDbContextResolver.cs]: Nope: will get rid of it

File: src/JsonApiDotNetCore/QueryParameterServices/SparseFieldsService.cs (hunk 2 above)
Comment (maurei) [reply src/JsonApiDotNetCore/QueryParameterServices/SparseFieldsService.cs]: will fix

File: test/JsonApiDotNetCoreExampleTests/Acceptance/Spec/SparseFieldSetTests.cs (hunk 4 above)
Comment (maurei) [reply test/JsonApiDotNetCoreExampleTests/Acceptance/Spec/SparseFieldSetTests.cs]: Yeah, agree. Fixing

**Pull request merged**